module/zfs/zvol.c

```diff
@@ -1012,6 +1021,97 @@ zvol_last_close(zvol_state_t *zv)
 	    !(zv->zv_flags & ZVOL_RDONLY))
 		txg_wait_synced(dmu_objset_pool(zv->zv_objset), 0);
 	(void) dmu_objset_evict_dbufs(zv->zv_objset);
+}
+
+/*
+ * Suspend the zvol for recv and rollback.
+ */
+int
+zvol_suspend(const char *name)
+{
+	zvol_state_t *zv;
+	int error = 0;
+
+retry:
+	mutex_enter(&zvol_state_lock);
+	zv = zvol_find_by_name(name);
+	if (zv == NULL) {
+		error = ENOENT;
+		goto out;
+	}
+
+	/* we cannot block in rw_enter if already suspended */
+	if (rw_owner(&zv->zv_suspend_lock) != NULL) {
```

Comment:
The issue here is we can't block while holding the `zvol_state_lock`, correct?

If so then this check doesn't completely resolve the possible deadlock.  The updated `zvol_request()` function takes the `zv->zv_suspend_lock` as a reader outside the `zvol_state_lock`.  This could happen before the write lock is taken below.  I think you're going to need to drop this check and rely on `rw_tryenter()`.

A little reference counting in the `zvol_state_t` so the `zvol_state_lock` could be dropped sooner wouldn't be a terrible thing either.